courses/views/v1/__init__.py

```diff
@@ -102,20 +104,54 @@ def paginate_queryset(self, queryset):
         return super().paginate_queryset(queryset)
 
 
+class CourseFilterSet(django_filters.FilterSet):
+
+    courserun_is_enrollable = django_filters.BooleanFilter(
+        field_name="courserun_is_enrollable",
+        method="filter_courserun_is_enrollable",
+    )
+
+    def filter_courserun_is_enrollable(self, queryset, _, value):
+        now = now_in_utc()
+
+        if value is True:
+            enrollable_runs = CourseRun.objects.filter(
+                Q(live=True)
+                & Q(start_date__isnull=False)
+                & Q(enrollment_start__lt=now)
+                & (Q(enrollment_end=None) | Q(enrollment_end__gt=now))
+            )
+            return queryset.prefetch_related(
+                Prefetch("courseruns", queryset=enrollable_runs)
+            ).filter(courseruns__id__in=enrollable_runs.values_list("id", flat=True))
+
+        else:
+            unenrollable_runs = CourseRun.objects.filter(
+                Q(live=False) | Q(start_date__isnull=True) | Q(enrollment_end__lte=now)
+            )
+            return queryset.prefetch_related(
+                Prefetch("courseruns", queryset=unenrollable_runs)
+            ).filter(courseruns__id__in=unenrollable_runs.values_list("id", flat=True))
+
+    class Meta:
+        model = Course
+        fields = ["id", "live", "readable_id", "page__live", "courserun_is_enrollable"]
+
+
 class CourseViewSet(viewsets.ReadOnlyModelViewSet):
     """API view set for Courses"""
 
     pagination_class = Pagination
     permission_classes = []
     filter_backends = [DjangoFilterBackend]
     serializer_class = CourseWithCourseRunsSerializer
-    filterset_fields = ["id", "live", "readable_id"]
+    filterset_class = CourseFilterSet
 
     def get_queryset(self):
         return (
             Course.objects.filter()
             .select_related("page")
-            .prefetch_related("courseruns", "departments")
```

Comment:
Should prefetching `courseruns` remain?  If we do decide to remove it, I noticed that a few n+1 warnings pop-up when calling the courses API.  
I was able to improve the query performance with the following changes:
1. Add `@cached_property` to `first_unexpired_run` on the `Course` model.
2. Remove `unexpired_runs` and `available_runs` from the `Course` model - these are not being used anywhere and could be replaced by filters in the future.
3. Change https://github.com/mitodl/mitxonline/blob/c09207714b4a728d8c7e2e613ed5c59b6c5b6d1c/courses/models.py#L491 to be `course_runs = self.courseruns.all().iterator()`.

Even after these changes I still notice one n+1 warning when performing GET: http://mitxonline.odl.local:8013/api/courses?id=1&live=true&page__live=true.

Reply:
I tested it with `courseruns` in prefetch_related here, there were two issues:
- It can't filter out any child `courserun` that already cached, it would just filters out results at the parent course level. For example, if we have a course with multiples runs, but only one of them meet criteria 3-6. With prefetch_related courseruns here, it would list all the runs for that course.
- To avoid the above issue, I added the prefetch_related `courserun` to `filter_courserun_is_enrollable` so we can filter out courseruns at the time of getting the object using Prefetch. But we can't have two prefetch_related courseruns, so I remove it from here.

I am wondering if it can be conditional based on the filter `courserun_is_enrollable` is applied or not. When `courserun_is_enrollable` isn't applied, add prefetch_related("courseruns"). I can check it out

Reply:
Ah,`prefetch_related("courseruns")` can be conditional based on `courserun_is_enrollable`. I put it back so when we call http://mitxonline.odl.local:8013/api/courses?id=1&live=true&page__live=true, it would use the prefetched courseruns object.

For these changes you mentioned, I can open a issue for that as I am hesitant to make these changes in this PR. These are good improvements to make, but I rather do some tests to make sure no side effects.

Reply:
Nice!